block-storage/src/main/scala/coop/rchain/blockstorage/dag/PersistentIndex.scala

```diff
@@ -0,0 +1,189 @@
+      result <- Sync[F].delay { byteBuffer.getLong() }.handleErrorWith {
+                 case _: BufferUnderflowException =>
+                   for {
+                     _ <- Log[F].warn(s"CRC file $crcPath did not contain a valid CRC value")
+                   } yield 0
+                 case exception =>
+                   Sync[F].raiseError(exception)
+               }
+    } yield result
+
+  private def calculateCrc[F[_]: Sync, K, V](
+      dataCollection: List[(K, V)],
+      keyValueCodec: Codec[(K, V)]
+  ): F[Crc32[F]] = {
+    val crc = Crc32.empty[F]()
+    dataCollection.traverse_[F, Unit] {
+      case (key, value) =>
+        keyValueCodec.encode((key, value)) match {
+          case Successful(encoded) =>
+            crc.update(encoded.toByteArray)
+          case Failure(e) =>
+            Sync[F].raiseError(EncodeError(e))
+        }
+    } >> crc.pure[F]
+  }
+
+  def readDataList[F[_]: Sync, K, V](
+      dataPath: Path,
+      keyValueCodec: Codec[(K, V)],
+      corruptedError: => StorageError
+  ): F[List[(K, V)]] = {
+    val inputStream = Files.newInputStream(dataPath, StandardOpenOption.READ)
+    val bitVector   = BitVector.fromInputStream(inputStream)
+
+    list(keyValueCodec).decode(bitVector) match {
+      case Successful(DecodeResult(keyValuePairList, BitVector.empty)) =>
+        keyValuePairList.pure[F]
+      case _ =>
+        Sync[F].raiseError(corruptedError)
+    }
+  }
+
+  private def validate[F[_]: Sync, K, V](
+      randomAccessIO: RandomAccessIO[F],
+      readCrc: Long,
+      dataList: List[(K, V)],
+      keyValueCodec: Codec[(K, V)],
+      corruptedError: => StorageError
+  ): F[(List[(K, V)], Crc32[F])] =
+    for {
+      fullCalculatedCrc <- calculateCrc(dataList, keyValueCodec)
+      result <- Monad[F].ifM(fullCalculatedCrc.value.map(_ == readCrc))(
+                 (dataList, fullCalculatedCrc).pure[F],
+                 if (dataList.isEmpty) {
+                   Sync[F].raiseError[(List[(K, V)], Crc32[F])](corruptedError)
+                 } else {
+                   // Trying to delete the last log entry which is most likely to be corrupted
+                   val dataWithoutLast = dataList.init
+                   calculateCrc(dataWithoutLast, keyValueCodec).flatMap { withoutLastCrc =>
+                     Monad[F].ifM(withoutLastCrc.value.map(_ == readCrc))(
+                       for {
+                         _ <- truncateDataLog(randomAccessIO, dataList, keyValueCodec)
+                       } yield (dataWithoutLast, withoutLastCrc),
+                       Sync[F].raiseError[(List[(K, V)], Crc32[F])](corruptedError)
+                     )
+                   }
+                 }
+               )
+    } yield result
+
+  def load[F[_]: Sync: Log: RaiseIOError, K: Codec, V: Codec](
+      logPath: Path,
+      crcPath: Path,
+      corruptedError: => StorageError
+  ): F[PersistentIndex[F, K, V]] = {
+    val keyValueCodec = ("key" | Codec[K]) ~ ("value" | Codec[V])
```

Comment:
shouldn't this be computed once?

Reply:
It is being computed only once for each specific index. Note that I can't move this computation outside of this method as it depends on specific implementations of `Codec[K]` and `Codec[V]`